### PR TITLE
Support AWS ECR credentials in China

### PR DESCRIPTION
**Pull request by zzq889**
**What this PR does / why we need it**:
Resolve AWS ECR credentials bug in China region

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

I haven't open any issue yet

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by therc** [comment]
Whoa, it's kinda crazy that they went for a separate domain, but I can see why. Is there any Amazon documentation? Even amazonaws.cn does not mention it.

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 1]
```diff
@@ -30,7 +30,9 @@ import (
 	"k8s.io/kubernetes/pkg/credentialprovider"
 )
 
+const chinaRegionPrefix = "cn-"
 const registryURLTemplate = "*.dkr.ecr.%s.amazonaws.com"
+const chinaRegistryURLTemplate = "*.dkr.ecr.%s.amazonaws.com.cn"
```
Comment (therc) [inline pkg/credentialprovider/aws/aws_credentials.go]: Is it amazonaws.com.cn or amazonaws.cn?

Comment (zzq889) [inline pkg/credentialprovider/aws/aws_credentials.go]: I'm sure is `amazonaws.com.cn`.

screen shot 2017-08-04 at 5 13 49 am

Comment (therc) [inline pkg/credentialprovider/aws/aws_credentials.go]: Thanks, it just looked weird, given that the site is on amazonaws.cn...

I kinda verified that at least the domain seems to resolve and point to a proxy to somewhere...
```

**Review by therc — APPROVED** [review]

File: pkg/credentialprovider/aws/aws_credentials.go (hunk 1 above)
Comment (zzq889) [inline pkg/credentialprovider/aws/aws_credentials.go]: amazonaws.cn is kind like aws.amazon.com for web visit and amazonaws.com.cn is the mirror site for amazonaws.com. It's really painful here to deal with the GFW related problems.

**Comment by zzq889** [comment]
Should I wait for the ok-to-test?

**Comment by zzq889** [comment]
@therc can you type ok-to-test?

**Comment by therc** [comment]
@k8s-bot ok to test

**Comment by therc** [comment]
@k8s-bot test this

**Comment by liggitt** [comment]
I'm not familiar enough with aws to know if this is reasonable to assume, or just happens to work (also, are there other geos this same change is needed for?). will defer to @kubernetes/sig-aws-misc

File: pkg/credentialprovider/aws/aws_credentials.go [hunk 2]
```diff
@@ -80,6 +82,14 @@ type ecrProvider struct {
 
 var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
 
+// registryURL has different suffix in AWS China region
+func registryURL(region string) string {
+	if strings.HasPrefix(region, chinaRegionPrefix) {
+		return fmt.Sprintf(registryURLTemplate, region)
+	}
+	return fmt.Sprintf(chinaRegistryURLTemplate, region)
```
Comment (therc) [inline pkg/credentialprovider/aws/aws_credentials.go]: These are swapped

**Comment by therc** [comment]
@liggitt this is reasonable to assume (I wrote the original ECR code). It shouldn't be needed for other regions. China is a special case, because of ICP licensing and so on. @zzq889 can you add a test?

**Review by therc — CHANGES_REQUESTED** [review]
You can add the .cn domain in otherRegistries in the aws_credentials_test.go:TestEcrProvide test. Even better, you can change the test to try verify that it works with both the existing lala-land-1 region and cn-foo-1.

**Comment by zzq889** [comment]
@therc Sure. I'm working on it.

**Comment by zzq889** [comment]
@liggitt depends on https://github.com/aws/aws-sdk-go/blob/b7eb6853a2d4bf9238b6613874d62006e0d6c27a/models/endpoints/endpoints.json aws currently has three kind of regions: `AWS Standard`, `AWS GovCloud (US)`, `AWS China`

**Comment by zzq889** [comment]
@therc I've update the code and testcase. Is ok-to-test still required for retest?

**Comment by zzq889** [comment]
@therc Even if I revert the code I still got test fail:
```
--- FAIL: TestEcrProvide (0.00s)
	aws_credentials_test.go:83: Didn't find expected URL: 123456789012.dkr.ecr.lala-land-1.amazonaws.com/foo/bar
FAIL
```
Any ideas?

**Comment by justinsb** [comment]
/lgtm

/retest

I also can't see the reason for the test failure, but if the restest doesn't get it I can fire it up in a debugger

**Comment by justinsb** [comment]
@zzq889 awsStandardDNSSuffix should be amazonaws.com:

```
-const awsStandardDNSSuffix = "amazon.com"
+const awsStandardDNSSuffix = "amazonaws.com"
```

**Comment by lareeth** [comment]
Any chance this can be added into a 1.6 release as well?

**Comment by lareeth** [comment]
@deads2k Any idea when this can be approved? This is super critical for running in China.

**Comment by deads2k** [comment]
> @deads2k Any idea when this can be approved? This is super critical for running in China.

I see the comment as outstanding: https://github.com/kubernetes/kubernetes/pull/50108#issuecomment-326473815

Also, needs a squash.

**Comment by lareeth** [comment]
@zzq889 When do you plan to make this fix? #50108

**Comment by lareeth** [comment]
@zzq889 Any update? Its been a month.

**Comment by zzq889** [comment]
I use this [container](https://github.com/mixslice/ecr-dockercfg-refresh) and wrote a [cronjob](https://github.com/mixslice/trident/blob/master/addons/ecr-dockercfg-refresh/ecr-dockercfg-refresh.yml) for replacement.

This pull request didn't pass the test with unknown reasons which may exist before I fork. I'll rebase and submit ag.

**Comment by zzq889** [comment]
```
I1115 08:42:08.488] process 23 exited with code 0 after 0.1m
I1115 08:42:08.489] Call:  git checkout -B test c3e4084066105ff632668deb3732cdb276b0e622
W1115 08:42:08.713] Reset branch 'test'
I1115 08:42:08.716] process 35 exited with code 0 after 0.0m
I1115 08:42:08.717] Call:  git merge --no-ff -m 'Merge +refs/pull/50108/head:refs/pr/50108' 65bed1d982c10b386ddcc3ea899501df448f8565
W1115 08:42:08.732] merge: 65bed1d982c10b386ddcc3ea899501df448f8565 - not something we can merge
E1115 08:42:08.733] Command failed
```
What I need to do with this error? @justinsb

**Comment by justinsb** [comment]
/retest

That just looks like a test flake

**Pull request merged**